### ISSUE TITLE
Issue pulling images with docker compose in version 24.0.0

### ISSUE THREAD

**Issue by mccaig** [docker/cli#4295]
### Description

Hi!

I installed docker on a new server this afternoon (ubuntu jazzy) and immediately started having issues with `docker compose pull`. It begins downloading the images in the compose file, but after a few seconds fails with `No such image: [imagename]`, even though the image exists.
After reverting to version 23.0.6 I have no issues.

### Reproduce

`docker compose -f [compose file with many images] pull`

### Expected behavior

docker should successfully pull images in the compose file

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.0
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        98fdcd7
 Built:             Mon May 15 18:49:22 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.0
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       1331b8c
  Built:            Mon May 15 18:49:22 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.4
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.17.3
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 24.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-1023-aws
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 30.68GiB
 Name: ip-10-251-201-19
 ID: 4b75e394-cbb2-42dd-aed4-cbf55fd925d2
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Perhaps as data-points as well;

- does it only reproduce with a compose file with many images?
- ^ is it possible to provide the compose file, or is it using private images that you can't share?
- are you able to collect logs of the docker daemon, in case it contains useful information?

**Comment by atesca09** [comment]
### docker version
`Client: Docker Engine - Community
 Version:           24.0.0
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        98fdcd7
 Built:             Mon May 15 18:48:35 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.0
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       1331b8c
  Built:            Mon May 15 18:48:35 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0`

### docker-compose version
`Docker Compose version v2.18.0`


### docker-compose.yml
`version: '3.9'

services:
  solr:
    build:
      context: ./solr/

  mysql:
    image: mysql:8.0
`

### ./solr/Dockerfile
`FROM solr:8.11`

### Command
`docker-compose build --pull`

Fails with `ERROR [solr internal] load metadata for docker.io/library/solr:8.11`
Without the `--pull` parameter it works just fine

**Comment by thaJeztah** [comment]
That last error seems like it's coming from BuildKit /cc @crazy-max

**Comment by atesca09** [comment]
You can ignore the error in my last comment. Purging the registry setup as a pull through cache completely, solved the issue for me.

**Comment by salob** [comment]
We also had this issue after upgrading docker from
old:
`Docker version 23.0.6, build ef23cbc`
new:
`Docker version 24.0.0, build 98fdcd7`

`docker compose version` is the same on both
`Docker Compose version v2.17.3`

Installed via yum on rocky

Downgrading docker to 23.0.6 fixes the issue.

Issue manifested for us as `No such image: nginx:latest` even though we were logged in to dockerhub and there was no `nginx:latest` reference in our docker-compose.yml file. In fact the correct `nginx:mainline-alpine` was pulled prior to the error. 

After initially believing upgrading docker compose plugin would resolve the issue, I upgraded to `v2.18.0`. However this time I go the same error but without any image details:
`Error: No such image:`

We are using multiple compose files i.e.

`docker compose -f docker-compose.run.yml -f docker-compose.dev.yml up -d --pull always`

I did try to just run:

`docker compose -f docker-compose.run.yml pull`

but got the same error.

We have another pipeline that uses a simpler single compose file that did not seem to be affected by the upgrade but I haven't done a full comparison to see what the main differences are between the compose yaml files yet.

**Comment by laurazard** [comment]
Right, as has been mentioned before, the `Error: No such image: ` on `Docker Compose version v2.18.0` is likely caused by https://github.com/docker/compose/issues/10574.

Other than that, I'm having trouble reproducing the `No such image: [imagename]` on `Docker Compose version v2.17.3` with `24.0.0` (I tried using a Compose file with `nginx:mainline-alpine` and some other services/images as mentioned by @salob to no effect).

If anyone can share a Compose file that reproduces this, or some daemon logs we'd really appreciate it.

**Comment by laurazard** [comment]
I made https://github.com/moby/moby/pull/45561 to address the issue in https://github.com/moby/moby/issues/45558, which seems similar to this – for confirmation, for the people who experiences this in this issue, are you specifying a platform in your Compose file?

**Comment by michel-laterman** [comment]
I've encountered this issue with `docker pull --platform= ...` without using compose

**Comment by mccaig** [comment]
The issue does seem to be related to the `platform` parameter.

This compose file will reproduce the error:
```
services:
  redis:
    image: redis:6
    platform: linux
```
What's interesting, is that the error message is for the `latest` tag:
`No such image: redis:latest`

specifying `latest` in the compose file, or alternatively removing the platform with avoid the issue, both of these compose files work for me:
```
services:
  redis:
    image: redis:6
```

```
services:
  redis:
    image: redis:latest
    platform: linux
```

Will try to fetch some logs from that daemon.

**Comment by laurazard** [comment]
Great, if it's all the same issue it's fixed by https://github.com/moby/moby/pull/45561 and we'll ship a moby `v24.0.1` with the fix soon. Thanks all 😊
